src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java

```diff
@@ -0,0 +1,470 @@
+        if (l.getPosition() > vc.getEnd())
+            return vc.getEnd() - l.getPosition();
+        return 0;
+    }
+
+    /**
+     * Parses a "Directive" of the form "ERROR,STRATIFIER,STRATIFIER...etc." into a {@link BaseErrorAggregation} consisting of the appropriate
+     * {@link BaseErrorCalculation.BaseCalculator BaseCalculator} and the successively "paired" {@link ReadBaseStratification.RecordAndOffsetStratifier RecordAndOffsetStratifier} as listed.
+     * The conversion from string to object is performed by the enums
+     * {@link BaseErrorCalculation.Errors Errors} and {@link ReadBaseStratification.Stratifiers Stratifiers}, and the "pairing" is performed by
+     * repeatedly wrapping two stratifiers in a {@link ReadBaseStratification.PairStratifier PairStratifier}
+     *
+     * @param stratifierDirective The string directive describing the error type and collection of stratifiers to use
+     * @return The appropriate {@link BaseErrorAggregation} object.
+     */
+    protected static BaseErrorAggregation parseDirective(final String stratifierDirective) {
+        final String[] directiveUnits = new String[256];
+        final char directiveSeparator = ':';
+        final int numberOfTerms = ParsingUtils.split(stratifierDirective, directiveUnits, directiveSeparator, false);
+
+        if (numberOfTerms == 256) {
+            throw new RuntimeException("Cannot parse more than 255 terms in a single directive. Sorry. (What are you doing?)");
+        }
+        if (numberOfTerms == 0) {
+            throw new RuntimeException("Found no directives at all. Cannot process.");
+        }
+
+        // make a linkedList due to removal and addition operations below
+        final List<ReadBaseStratification.RecordAndOffsetStratifier<?>> stratifiers = new LinkedList<>(Arrays.stream(directiveUnits)
+                .limit(numberOfTerms)
+                .skip(1)
+                .map(String::trim)
+                .map(ReadBaseStratification.Stratifiers::valueOf)
+                .map(ReadBaseStratification.Stratifiers::makeStratifier)
+                .collect(Collectors.toList()));
+
+        final ReadBaseStratification.RecordAndOffsetStratifier<? extends Comparable> jointStratifier;
+
+        if (stratifiers.isEmpty()) {
+            jointStratifier = ReadBaseStratification.nonStratifier;
+        } else {
+            // Repeatedly, remove the first two stratifiers in the list, wrap them in a PairStratifier
+            // and put it back in the list (in the front). Stop when there is exactly 1 element in the list.
+            while (stratifiers.size() > 1) {
+                ReadBaseStratification.RecordAndOffsetStratifier<?> first = stratifiers.remove(0);
+                ReadBaseStratification.RecordAndOffsetStratifier<?> second = stratifiers.remove(0);
+                ReadBaseStratification.PairStratifier<? extends Comparable, ? extends Comparable> newPair = new ReadBaseStratification.PairStratifier<>(first, second);
```

Comment:
Maybe it would've been better to have an N-Stratifier in addition to Pair?

Reply:
that would have made the signature of the class very confusing (I actually don't know how to do it)

Reply:
Hmm, the only problem I see is that `Collection` isn't `Comparable`. There's probably a way to wrap this but I don't know it offhand

```
    public static class CollectionStratifier<T extends Comparable<T>> implements RecordAndOffsetStratifier<Collection<T>> {
        final Collection<RecordAndOffsetStratifier<T>> stratifiers;

        public CollectionStratifier(Collection<RecordAndOffsetStratifier<T>> stratifiers) {
            this.stratifiers = stratifiers;
        }

        @Override
        public Collection<T> stratify(RecordAndOffset recordAndOffset, SAMLocusAndReference locusInfo) {
            return stratifiers.stream().map(o -> o.stratify(recordAndOffset, locusInfo)).collect(Collectors.toList());
        }

        @Override
        public String getSuffix() {
            return stratifiers.stream().map(RecordAndOffsetStratifier::getSuffix).collect(Collectors.joining("_and_"));
        }
    }
```

Reply:
The problem I was refering to is that in the Pair, left and right can have different types...

Reply:
> The problem I was refering to is that in the Pair, left and right can have different types

As far as I can tell, in this code fragment the type argument to `Pair` is not bound. This is because the type is generated by a series of calls to `Stratifiers.makeStratifier()`, and the result of that operation is `RecordAndOffsetStratifier<?>`

Reply:
I see what you mean. I'll add a CollectionStratifier that will remove this code from the CLP.